### PR TITLE
CI/CD basic implementation based on GitHub Actions

### PR DESCRIPTION
**Pull request by ia**
TL;DR: utilize GitHub capabilities to make builds on every commit and on every pull-request.

Hello. I'm a bit surprised that this haven't been done yet. So I decided to take some liberty to help with that. There may be a lot of room for improvements in the future but this is _work-able_ proof-of-concept which can be merged & adopted as is currently in my humble opinion. However, to fully integrate `push.yml` into `hydrafw` github project, admins may need to set some additional settings _somewhere_ in _Settings_ of the project using GitHub web interface manually.

The benefits are:
- always be sure that current code at least _build-able_;
- without need to enroll local devel environment, advanced users who just will need `dfu-util`, always be able to get the latest fresh build from main branch;

Partially this overlaps with #132 and I can't promise anything but every time I have some free time I try to keep contributing around these CI/CD improvements as well.

Now, to technical details so everyone who is interested & curious could be on the same page. Basically, `push.yml` here is the most minimal declarative file which tells _GitHub_ to start _Ubuntu_ `22.04` VM, then to start _Debian_ `oldstable` container inside, install toolchain packages, build `hydrafw.dfu` and upload it as `hydrafw.dfu` artifact. The live _"demo"_ can be seen in my forked repo [here](https://github.com/ia/hydrafw/actions):
 - [overview](https://github.com/ia/hydrafw/actions)
 - [latest green build](https://github.com/ia/hydrafw/actions/runs/8401295543)
 - [artifact section](https://github.com/ia/hydrafw/actions/runs/8401295543#artifacts)
 - [direct link to `hydrafw.zip`](https://github.com/ia/hydrafw/actions/runs/8401295543/artifacts/1352346834)
 
 Now, why such "matreshka" with VMs/containers, you may ask. Well, _GitHub_ doesn't have _Debian_ `oldstable` to run it directly, but we can start it from VM which is available on _GitHub_. After tries & errors _Debian_ `oldstable` has been picked as [so far] the most suitable one, since:
 - it has `git` with fresh enough version to support `actions/checkout@v4` (it needs `git` > 2.17)
 - it has old enough `GCC` toolchain where _LTO_ for `ChibiOS` doesn't have [a strange issue](http://forum.chibios.org/viewtopic.php?t=5830) ([sample log from one of my builds](https://github.com/ia/hydrafw/actions/runs/8388352807/job/22972355589#step:8:202))
 
 If you have any other questions, just let me know.
 
Oh, and that tiny update for `python` script comes from specific behavior of _GitHub_ actions: on every checkout before build, _GitHub_ infrastructure doesn't make full valid `git checkout` but downloads source tree in detached state so getting meta information about repo and its status (tag, revision, sha, etc.) may be tricky because I saw that with other GitHub projects so for a while we set revision as `0.0` for _GitHub_ builds (which is default value) but it's even good in its own way because by that it will be the indicator that it is _the latest upstream mainline build_ but not a stable release.

### CODE REVIEW

**Comment by ia** [comment]
[Aaaand off we go!](https://github.com/hydrabus/hydrafw/actions)
So it seems it's picked automagically even with this pull-request right away for the main hydrabus/hydrafw project:
- [CI tab](https://github.com/hydrabus/hydrafw/actions/workflows/push.yml)
- [build page](https://github.com/hydrabus/hydrafw/actions/runs/8435005160)
- [zip artifact](https://github.com/hydrabus/hydrafw/actions/runs/8435005160#artifacts) and [direct link to it](https://github.com/hydrabus/hydrafw/actions/runs/8435005160/artifacts/1358955461)

**Comment by bvernoux** [comment]
Thanks for the contribution but could you apply the official tested & validated with [GCC ARM 4.9 2015q3(GNU_ARM_4_9_2015q3)](https://launchpad.net/gcc-arm-embedded/+milestone/4.9-2015-q3-update) used and defined in
https://github.com/hydrabus/hydrafw/wiki/how-to-build-flash-and-use-hydrafw-on-linux ?
- Like that I could cross check that the dfu created is exactly what I have validated during all those years (using the old [GCC ARM 4.9 2015q3(GNU_ARM_4_9_2015q3)](https://launchpad.net/gcc-arm-embedded/+milestone/4.9-2015-q3-update) which works perfectly as newer version does not bring anything except some regression risks ...)

**Comment by ia** [comment]
> could you use the official tested & validated [GCC ARM 4.9 2015q3(GNU_ARM_4_9_2015q3)](https://launchpad.net/gcc-arm-embedded/+milestone/4.9-2015-q3-update) used and defined in
> https://github.com/hydrabus/hydrafw/wiki/how-to-build-flash-and-use-hydrafw-on-linux ?

Sure! Thanks for pointing this out. I perfectly understand that for specific hardware projects it may be _very essential_ to have reference devel environment to avoid a chance of wasting time to deal with toolchain problem instead of improving the project itself. I just forgot about that `GCC` version requirement for `hydrafw` since I didn't have any issues myself with host `arm-none-eabi` toolchain on `18.04` [yet].

Basically, I just need to think how better incorporate downloading, unpacking & setting that toolchain. Will try to be back with results today or tomorrow.

**Comment by ia** [comment]
After very brief look, it looks good to me.
And [these](https://github.com/hydrabus/hydrafw/actions/runs/8449573440/job/23143930307#step:8:7) [two](https://github.com/hydrabus/hydrafw/actions/runs/8449573440/job/23143930307#step:8:13) log lines give a hint that now the reference toolchain is used.
And here is [the link](https://github.com/hydrabus/hydrafw/actions/runs/8449573440) to the latest successful build, in case if you (as far as I understand) is interested in some kind of _reproducible build_ analysis of the resulting `dfu` file ([direct link](https://github.com/hydrabus/hydrafw/actions/runs/8449573440/artifacts/1362399528) to `zip` with `hydrafw.dfu` after setting toolchain).

P.S. For the record: I'm not a big fan of modern Dev[Put Something Else Fancy Today Here]Ops over-hyped buzzwords but even that may be useful with reasonable approach. And what I really like in other Free & Open source projects is that when you can just go to download/build section/tab and download the latest build/app without a need to spend time to set up devel environment yourself, especially if you just want to try/test something "right here and right now" and you're in a rush. So since `hydrafw` is located on GitHub and since GitHub provides such infrastructure for free (as in beer), why wouldn't utilize it for the benefits of the project, right? ;)

**Comment by bvernoux** [comment]
I have checked your latest build https://github.com/hydrabus/hydrafw/actions/runs/8449573440/artifacts/1362399528 and it seems there is an issue in the Firmware version embedded in the DFU `HydraFW (HydraBus) 1738895 2024-03-27` instead of expected `HydraFW (HydraBus) v0.11-22-g34e3d63 2024-02-23`

File: .github/workflows/push.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+      - name: Install packages
+        run: sudo  apt  install  -y  --no-install-recommends  --no-install-suggests  bash  coreutils  tar  bzip2  git  make  python3  python3-pip
+
+      - name: Install python modules
+        run: python3  -m pip  install  GitPython intelhex  --upgrade
+
+      - uses: actions/checkout@v4
+        with:
+          submodules: true
```
Comment (Baldanos) [inline .github/workflows/push.yml:33]: Does adding `fetch-tags: true` allows to run the Python script without errors ?

Comment (ia) [reply .github/workflows/push.yml:33]: [Nope](https://github.com/hydrabus/hydrafw/actions/runs/8456176252/job/23165330015#step:8:15). Maybe it's because what is built here is not from specific release tag but from main devel branch with commit not related to any specific release tag.

Comment (weakfl) [reply .github/workflows/push.yml:33]: That's a bug in the v4 checkout action. Currently afk but you should find details in the respective repo (including workarounds).

**Comment by ia** [comment]
> I have checked your latest build https://github.com/hydrabus/hydrafw/actions/runs/8449573440/artifacts/1362399528 and it seems there is an issue in the Firmware version embedded in the DFU `HydraFW (HydraBus) 1738895 2024-03-27` instead of expected `HydraFW (HydraBus) v0.11-22-g34e3d63 2024-02-23`

Aha. Well, I may be wrong but probably I was not so clear presenting this feature. So what you're looking for in `hydrafw.dfu` is a _version string_. But this string can be seen only in binary which is built from git release tag with `v0.11` value and with related to that tag git sha1 commit id which is `g34e3d63`. But builds on GitHub actions are "baked" by checking out the most recent commit from main upstream devel branch (which is `master`) so _version string_ should be different depends on commit.

However, the mystery is why `1738895`? It seems I need more time to figure out what's what and how exactly this string is generated diving a bit deeper into related python scripts. But I appreciate the interest to this work. Now we all need to agree on some kind of _version string_ format which will be indicative to determine tag/branch/commit_id but at the same time won't be pain in the neck to get this meta information during build, whether local or github or from tarball (where git meta info is not available at all). My suggestion is something like: `HydraFW (HydraBus) GIT_TAG_if_available GIT_SHA_ID_if_availablee BUILD_DATE`. But this is just for reference. So just tell me how it should be and I will try to figure something out.

**Comment by ia** [comment]
> > > I have checked your latest build https://github.com/hydrabus/hydrafw/actions/runs/8449573440/artifacts/1362399528 and it seems there is an issue in the Firmware version embedded in the DFU `HydraFW (HydraBus) 1738895 2024-03-27` instead of expected `HydraFW (HydraBus) v0.11-22-g34e3d63 2024-02-23`
> > 
> > 
> > Aha. Well, I may be wrong but probably I was not so clear presenting this feature. So what you're looking for in `hydrafw.dfu` is a _version string_. But this string can be seen only in binary which is built from git release tag with `v0.11` value and with related to that tag git sha1 commit id which is `g34e3d63`. But builds on GitHub actions are "baked" by checking out the most recent commit from main upstream devel branch (which is `master`) so _version string_ should be different depends on commit.
> > However, the mystery is why `1738895`? It seems I need more time to figure out what's what and how exactly this string is generated diving a bit deeper into related python scripts. But I appreciate the interest to this work. Now we all need to agree on some kind of _version string_ format which will be indicative to determine tag/branch/commit_id but at the same time won't be pain in the neck to get this meta information during build, whether local or github or from tarball (where git meta info is not available at all). My suggestion is something like: `HydraFW (HydraBus) GIT_TAG_if_available GIT_SHA_ID_if_availablee BUILD_DATE`. But this is just for reference. So just tell me how it should be and I will try to figure something out.
> 
> I really prefer to keep it like it is today with Release using latest release version "v0.11" (or newer) then use number of new commits since that official release

No problem. Maybe I will figure out eventually how to do that, maybe not because this is very tricky version string policy. I mean, I've just never heard about counting commits.

But now I'm just really very curious to understand... because the more it will be clear to me, the more chances that I will do it as demanded without making some mistakes because something unclear to me. So...

> see https://github.com/hydrabus/hydrafw/releases/tag/v0.11 today we have 22 commits after v0.11 so it shall be formatted v0.11-22

... correct me if I'm wrong but as far as I understand, `v0.11` is just a git tag, and git tag is just a human readable alias with the pointer to the specific commit _in the past_. Once any commit is committed on top of the tagged commit, tagged commit is losing it's "recent" status and become out-of-dated by definition. But here it should be not only kept but on top of that there is counting of number of commits between release tag and the most recent commit on top of upstream devel branch... why so complicated? Git and its commit id system allows to establish exact source tree state from which build has been produced just by sha id. Tags can be embedded if it's build from commit with the exact tagged sha id. I'm just definitely missing something here which I really try to understand, sorry to bother with these questions.

UPD. Okay, I see that this is some sort of utilization of `git  describe  --tags`... but why they would add commit counter in the first place... interesting. Anyway, you see, another one tricky part: if build is produced from pull-request [feature] branch like here from my `ia:master` branch where I already currently have ~50 commits, then what version string should be for the builds from `ia:master` here:

a) `v0.11-73-g595a6e1` (which is technically valid but not correct from the look of `hydrafw:master` especially when after merge `73` turns into `23`);

b) `v0.11-22-g595a6e1` (which is correct from the look of `hydrafw:master` but tag `595a6e1` originally corresponds to commit inside feature branch and not inside `hydrafw:master` which may be a bit confusing in cases of searching for regression or triaging an issue);

c) `v0.11-22-g34e3d63` (which makes the source tree state untraceable at all because the meta information doesn't correspond to the actual source tree state at the time of build because this version string won't be pointing out to the source tree state at the time of build which is the whole point of version string as far as I understand);

?

P.S. Regardless, today I learned so much already. Much appreciate the feedback. :pray:

File: .github/workflows/push.yml (hunk 1 above)
Comment (ia) [reply .github/workflows/push.yml:33]: > That's a bug in the v4 checkout action. Currently afk but you should find details in the respective repo (including workarounds).

Yea... that whole `v4` thing brought a lot of headache to a lot of people. For example, with `upload v4` update they disabled ability to upload multiple artifacts from multiple matrix'es into the zip with the same name.

---

Anyway, to whom it may concern and who is just curious: _it seems_ `fetch-depth: 0` helps to fetch tags and _some_ meta information but not _branches_ meta info (which is needed to count `22` commits) because `fetch-...` options don't change the fact that the tree itself ends up on github server in detached state right before the build.

**Comment by ia** [comment]
Ok, _it seems_ that after all with the power of bash & git commands we can get all needed meta information at the time of building on github to embed _version string_.

The question remains what _string_ should be put into `hydrafw.dfu`, so considering my comment above just let me know and I will do some further modifications. I hope my comments are not looking rude in any way because I was just trying to figure things out. Will return to this in a day or too.

**Comment by bvernoux** [comment]
> Ok, _it seems_ that after all with the power of bash & git commands we can get all needed meta information at the time of building on github to embed _version string_.
> 
> The question remains what _string_ should be put into `hydrafw.dfu`, so considering my comment above just let me know and I will do some further modifications. I hope my comments are not looking rude in any way because I was just trying to figure things out. Will return to this in a day or too.

We must just build it like it was done on computer it create in src/common/hydrafw_version.hdr which is then used during build and those informations will be present in the final dfu (like it is in the bin ...) so there is nothing more to do 
Example content of src/common/hydrafw_version.hdr created during build with latest commit:
```
#define HYDRAFW_GIT_TAG "v0.11-22-g34e3d63"
#define HYDRAFW_CHECKIN_DATE "2024-02-23"
```

File: src/build-scripts/hydrafw-version.py [hunk 2]
```diff
@@ -14,7 +16,12 @@
 if len(args)==1:
   sys.stdout = open(args[0], 'w')
   git=Repo(search_parent_directories=True).git
-  print('#define HYDRAFW_GIT_TAG "' + git.describe(tags=True,always=True,dirty=True,long=True) + '"')
+  vermagic=git.describe(tags=True,always=True,dirty=True,long=True)
+  if os.environ.get("GITHUB_CI_PR_SHA", "") != "":
+    sha_id=os.environ["GITHUB_CI_PR_SHA"][:7]
+    print('#define HYDRAFW_GIT_TAG "' + re.sub('-g.*$', '-g' + sha_id, vermagic) + '"')
+  else:
+    print('#define HYDRAFW_GIT_TAG "' + vermagic + '"')
```
Comment (ia) [inline src/build-scripts/hydrafw-version.py:24]: I know that on the long run the goal is to remove python dependency at all and not to extend it even deeper, but currently I have to work with what we have. Which is adopting of extraction proper git tag at the moment of build on GitHub server. Although after helpful conversations & recommendations I did manage to push through all necessary git meta information for the build, one essential GitHub build issue still remains: the repo is in detached state at the time of build, so we can't get valid the most recent committed sha id easily. Therefore I had to use already tested approach: we set env variable in `push.yml` with actual sha id since GitHub allows to extract this information at the time of build. And then during build we check if sha id env is set, and then we replace detached sha id by actual one. I guess it's self-explanatory anyway but just for those who may be curious.

Comment (bvernoux) [reply src/build-scripts/hydrafw-version.py:24]: All seems good when checking the latest hydrafw.dfu
Do you plan to do some cleanup for example in scripts/env.sh ... (before to merge it) ?

Comment (ia) [reply src/build-scripts/hydrafw-version.py:24]: > All seems good when checking the latest hydrafw.dfu

Perfecto!

> Do you plan to do some cleanup for example in scripts/env.sh ... (before to merge it) > ?

Yes-yes, sure, of course! Just give me a day or so since I plan to make a couple more commits here, and then I will tell you when it's ready. Thanks.

File: .github/workflows/push.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+          fetch-depth: 0
+
+      - name: Set git meta info
+        run: echo "GITHUB_CI_PR_SHA=${{github.event.pull_request.head.sha}}" >> "${GITHUB_ENV}"
+
+      - name: Install standalone reference GCC toolchain
+        run: bash  scripts/env.sh
+
+      - name: Build ${{ matrix.model }}
+        run: source  build.env  &&  arm-none-eabi-gcc --version  &&  arm-none-eabi-gcc -print-search-dirs  &&  python3  src/build-scripts/hydrafw-revision.py  &&  make  V=1  -j$(nproc)  -C src/
```
Comment (ia) [inline .github/workflows/push.yml:44]: `python3 src/build-scripts/hydrafw-revision.py`
 <<< oops, missed that one to remove since it was for temp debugging as well...

File: src/build-scripts/hydrafw-version.py (hunk 2 above)
Comment (ia) [reply src/build-scripts/hydrafw-version.py]: Oookay, I guess now it should be ok. So consider it done on my side for now. But if you have any other additional suggestions or any feedback, just let me know.

P.S. I'm sincerely sorry if some of my questions or explanations were a bit intimidating or irritating here in the conversations ... but I was just trying to be straight-forward and direct only to make sure that I understand clearly what exactly must be implemented for this PR since build scripts & steps for `hydrafw` in the context of GitHub Actions environment turned out to be a little bit more tricky than I thought at first look.

**Comment by ia** [comment]
Hello. Sorry to bother, but I'm just giving a friendly reminder of [the most recent update](https://github.com/hydrabus/hydrafw/pull/167#discussion_r1547714473) since I'm very curious what you think.

So, could you, please, just give me the feedback about the current patch-set for this pull-request. Although I understand that we all may be very busy sometimes so take your time, of course, so I'm not rushing, but just giving an update. Thank you.

**Comment by bvernoux** [comment]
All seems good for me
Baldanos what do you think ?

**Comment by Baldanos** [comment]
Looks good to me.

@ia Thank you for the contribution !

**Comment by bvernoux** [comment]
Thanks for your contribution
